libvast/vast/system/source.hpp

```diff
@@ -257,46 +260,26 @@ source(caf::stateful_actor<source_state<Reader>>* self, Reader reader,
                                             push_slice);
       t.stop(produced);
       st.count += produced;
-      auto force_emit_batches = [&] {
-        st.mgr->out().fan_out_flush();
-        st.mgr->out().force_emit_batches();
-        st.send_report();
-      };
       auto finish = [&] {
         st.done = true;
         st.send_report();
         self->quit();
       };
-      if (produced == 0) {
-        // If the source is unable to generate new events (returns 0), the
-        // source will stall and never be polled again, because CAF expects
-        // stream sources to _always_ be able to generate new events. As a
-        // workaround, we send out an invalid table slice here that gets ignored
-        // downstream, and forcefully emit batches.
-        VAST_VERBOSE(self, "emits invalid table slice to prevent stalling");
-        out.push(nullptr);
-        return finish();
-      }
       if (st.requested && st.count >= *st.requested)
         return finish();
-      if (err != caf::none) {
-        if (err == vast::ec::timeout) {
-          if (produced > 0) {
-            VAST_DEBUG(self, "hit input timeout and forcefully emits", produced,
-                       "produced events");
-            return force_emit_batches();
-          } else {
-            // This case should never happen. If it does, we hit an internal
-            // application logic error in the reader. CAF might stall out on us
-            // here.
-            VAST_ERROR(self, "hit input timeout, but produced no events");
-            return finish();
-          }
-        } else {
-          if (err != vast::ec::end_of_input)
-            VAST_INFO(self, "completed with message:", render(err));
-          return finish();
+      if (err == ec::timeout) {
+        if (!st.waiting_for_input) {
+          // This pull handler was invoked while we were waiting for a wakeup
+          // message. Sending another one would create a parallel wakeup cycle.
+          st.waiting_for_input = true;
+          self->delayed_send(self, std::chrono::seconds{1}, atom::wakeup_v);
```

Comment:
Shouldn't this just be a normal send to enqueue the message to the back of the message queue?

Reply:
We agreed to implement an exponential backoff mechanism here.